src/Input.cpp

```diff
@@ -79,15 +81,22 @@ int Input::JoystickFromGUIDString(const char *guid)
 int Input::JoystickFromGUID(SDL_JoystickGUID guid)
 {
 	const int guidLength = 16; // as defined
-	for (auto pair : m_joysticks) {
-		JoystickInfo &state = pair.second;
+	for (unsigned i = 0; i < m_joysticks.size(); ++i) {
+		auto &state = m_joysticks[i];
 		if (0 == memcmp(state.guid.data, guid.data, guidLength)) {
-			return static_cast<int>(pair.first);
+			return i;
 		}
 	}
 	return -1;
 }
 
+int Input::JoystickFromID(SDL_JoystickID id)
+{
+	assert(id >= 0 && "this ID means no joystick");
+	assert(static_cast<unsigned>(id) < m_joyIDs.size() && "there is no internal ID for this SDL ID");
+	return m_joyIDs[id];
```

Comment:
@Web-eWorks
The fact is that values that do not satisfy the assertion are invalid states, a normally working program simply should not call this function with invalid parameters. Also I would like to note that the result of this function in 6 places is directly used to index the array, you suggest to write check in each of these places?
```
Input.cpp:
629: if (!GetJoysticks()[Input::JoystickFromID(event.jaxis.which)].joystick)
631: auto &axis = GetJoysticks()[Input::JoystickFromID(event.jaxis.which)].axes[event.jaxis.axis];
640: if (!GetJoysticks()[Input::JoystickFromID(event.jaxis.which)].joystick)
642: GetJoysticks()[Input::JoystickFromID(event.jbutton.which)].buttons[event.jbutton.button] = event.jbutton.state != 0;
645: if (!GetJoysticks()[Input::JoystickFromID(event.jaxis.which)].joystick)
647: GetJoysticks()[Input::JoystickFromID(event.jhat.which)].hats[event.jhat.hat] = event.jhat.value;
```

Reply:
Exactly half of the places that call this function are calling it redundantly to retrieve a value that hasn't changed since the first time it was called, so that can be cleaned up eventually. :smile:

As this is public API that has no guarantee whether the caller intends an invalid joystick ID to cause a program crash, I'd *vastly* prefer that it return an invalid value. The existing callsites do not need to check the return value as they have the external precondition that the joystick IDs are valid, as they're coming from SDL for joysticks that the program has explicitly opened and added to the array.

On the other hand, if this is made a private function and not public API I don't really see the value of having any checks at all here if it's not going to be called from outside the Input class, as we're able to completely enforce the precondtion of a valid SDL_JoystickID externally to the function and don't need the overhead of an assert.

Reply:
Having argued the point to death on IRC, I withdraw my recommendation as at current there's nothing using it where the assert would be particularly harmful.